pkg/portutil/port_allocate_linux.go

```diff
@@ -0,0 +1,125 @@
+//go:build linux
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package portutil
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"strings"
+)
+
+const (
+	tcpProto = "tcp"
+	udpProto = "udp"
+	// FIXME: The /proc/net/tcp is not recommended by the kernel, FYI https://www.kernel.org/doc/Documentation/networking/proc_net_tcp.txt
+	// In the future, we should use netlink instead of /proc/net/tcp
+	netTcpStats = "/proc/net/tcp"
+	netUdpStats = "/proc/net/udp"
+	// This port range is compatible with Docker, FYI https://github.com/moby/moby/blob/master/libnetwork/portallocator/portallocator_unix.go#L7-L12
+	allocateStart = 49153
+	allocateEnd   = 60999
+)
+
+func readStatsFileData(protocol string) ([]string, error) {
+	var fileAddress string
+
+	if protocol == tcpProto {
+		fileAddress = netTcpStats
+	} else if protocol == udpProto {
+		fileAddress = netUdpStats
+	} else {
+		return nil, fmt.Errorf("unsupported protocol %s", protocol)
+	}
+
+	data, err := ioutil.ReadFile(fileAddress)
+	if err != nil {
+		return nil, err
+	}
+	lines := strings.Split(string(data), "\n")
+
+	return lines[1 : len(lines)-1], nil
+}
+
+func removeEmpty(array []string) []string {
+	var results []string
+	for _, i := range array {
+		if i != "" {
+			results = append(results, i)
+		}
+	}
+	return results
+}
+
+func parseHex(h string) (uint64, error) {
+	port, err := strconv.ParseInt(h, 16, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(port), nil
+}
+
+func getUsedPort(proto string) (map[uint64]uint8, error) {
```

Comment:
This should take the content of /proc/net/* as `[]byte` argument, and should have unit tests

Reply:
FYI: you may take some code from https://github.com/lima-vm/lima/tree/v0.8.3/pkg/guestagent/procnettcp (Apache License 2.0)

Reply:
Looks good to me! Thx for your advice